File: core/contractsdk/java/example/game_assets/src/main/java/com/baidu/xuper/GameAssets.java

package com.baidu.xuper;

public class GameAssets implements Contract {
    final String ASSETTYPE = "AssetType_";
    final String USERASSET = "UserAsset";
    final String ASSET2USER = "Asset2User_";
    final String ADMIN = "admin";


    @Override
    @ContractMethod
    public Response initialize(Context ctx) {
        String admin = ctx.caller();
        if (admin == null || admin.isEmpty()) {
            return Response.error("missing admin address");
        }
        ctx.putObject(ADMIN.getBytes(), admin.getBytes());
        return Response.ok("ok".getBytes());
    }

    private boolean isAdmin(Context ctx) {
        return this.isAdmin(ctx, ctx.caller());
    }

    private boolean isAdmin(Context ctx, String caller) {
        String admin = new String(ctx.getObject("admin".getBytes()));
        return admin.equals(caller);
    }

    @ContractMethod
    public Response addAssetType(Context ctx) {
        if (!this.isAdmin(ctx)) {
            return Response.error("you do not have permission to call this method");
        }
        if (ctx.args().get("type_id") == null || ctx.args().get("type_desc") == null) {
            return Response.error("missing type_id or type_desc");
        }

        String typeId = new String(ctx.args().get("type_id"));
        String typeDesc = new String(ctx.args().get("type_desc"));
        if (typeId.isEmpty() || typeDesc.isEmpty()) {
            return Response.error("missing type_id or type_desc");
        }

        String assetTypeKey = ASSETTYPE + typeId;
        if (ctx.getObject(assetTypeKey.getBytes()) != null) {
            return Response.error("asset type "+typeId+" already exists");
        }
        ctx.putObject(assetTypeKey.getBytes(), typeDesc.getBytes());
        return Response.ok(typeId.getBytes());
    }

    @ContractMethod
    public Response listAssetType(Context ctx) {
        StringBuffer buf = new StringBuffer();
        ctx.newIterator(ASSETTYPE.getBytes(), (ASSETTYPE + "~").getBytes())
                .forEachRemaining(
                        elem -> {
                            String assetType = new String((elem.getKey()));
                            buf.append((assetType.substring(ASSETTYPE.length()) + ":" +
                                    new String(elem.getValue()) +
                                    "\n"));
                        }
                );
        return Response.ok(buf.toString().getBytes());
    }

    @ContractMethod
    public Response getAssetByUser(Context ctx) {
        if (ctx.args().get("user_id")==null){
            return Response.error("missing user_id");
        }
        String userId = new String(ctx.args().get("user_id"));
        String userAsstKey = USERASSET + userId + "_";
        String start = userAsstKey;
        String end = start + "~";
        StringBuffer buf = new StringBuffer();
        ctx.log("1");
        ctx.newIterator(start.getBytes(), end.getBytes()).forEachRemaining(
                elem -> {
                    String assetId = new String(elem.getKey());
                    String typeId = new String(elem.getValue());
                    String assetTypeKey = ASSETTYPE + typeId;
                    String assetDesc = new String(ctx.getObject(assetTypeKey.getBytes()));
                    buf.append("assetId=");
                    buf.append(assetId.split("_")[1]);
                    buf.append(",type.id=");
                    buf.append(typeId);
                    buf.append(",assetDesc=");
                    buf.append(assetDesc);
                    buf.append("\n");
                }
        );
        return Response.ok(buf.toString().getBytes());
    }

    @ContractMethod
    public Response newAssetToUser(Context ctx) {

        if (!this.isAdmin(ctx)) {
            return Response.error("you do not have permission to call this method");
        }
        if (ctx.args().get("user_id") == null ||
                ctx.args().get("type_id") == null ||
                ctx.args().get("asset_id") == null) {
            return Response.error("missing user_id or type_id or asset_id");
        }


        String userId = new String(ctx.args().get("user_id"));
        String typeId = new String(ctx.args().get("type_id"));
        String assetId = new String(ctx.args().get("asset_id"));
        if (userId.isEmpty() || typeId.isEmpty() || assetId.isEmpty()) {
            return Response.error("missing user_id or type_id or asst_id");
        }

        String assetKey = ASSET2USER + assetId;

        if (ctx.getObject(assetKey.getBytes()) != null) {
            return Response.error("asset " + assetId + " exists");
        }

        byte[] assetDesc = ctx.getObject((ASSETTYPE + typeId).getBytes());
        if (assetDesc == null || assetDesc.length == 0) {
            return Response.error("asset type " + typeId + " not found");
        }

        String userAssetKey = USERASSET + userId + "_" + assetId;

        ctx.putObject(userAssetKey.getBytes(), typeId.getBytes());

        ctx.putObject(assetKey.getBytes(), userId.getBytes());
        return Response.ok(assetId.getBytes());
    }

    public Response tradeAsset(Context ctx) {
        String from = ctx.caller();
        if (from.length() == 0) {
            return Response.error("missing initiator");
        }
        byte[] to = ctx.args().get("to");
        String assetId = new String(ctx.args().get("assetid"));
        if (to.length == 0 || assetId.length() == 0) {
            return Response.error("missing to or assetid");
        }
        String userAssetKey = USERASSET + from + "_" + assetId;
        String assetType = new String(ctx.getObject(userAssetKey.getBytes()));
        if (assetType.length() == 0) {
            return Response.error("you do not have asset with id " + assetId);
        }

        ctx.deleteObject(userAssetKey.getBytes());
        String assetKey = ASSET2USER + assetId;
        String newUserAssetKey = USERASSET + to + "_" + assetId;
        ctx.putObject(newUserAssetKey.getBytes(), assetId.getBytes());
        ctx.putObject(assetKey.getBytes(), to);
        return Response.ok(assetId.getBytes());
    }

    public static void main(String[] args) {
        Driver.serve(new GameAssets());
    }
}

File: core/contractsdk/cpp/example/game_assets.cc
#include <sstream>

#include "xchain/xchain.h"

// 游戏装备资产模板
// 参数由xchain::Contract中的context提供
class GameAssets {
public:
    /*
     * func: 初始化装备发行账户
     * @param: admin: 哪个address具有管理员权限
     */
    virtual void initialize() = 0;
    /*
     * func: 新增装备类型
     * @param: initiator:交易发起者,只有交易发起者等于admin账户才能执行成功
     * @param: type_id: 游戏状态的参数和属性描述
     * @param: typedesc: 游戏状态的参数和属性描述
     */
    virtual void addAssetType() = 0;
    /*
     * func: 获取所有的装备类型和参数信息
     */
    virtual void listAssetType() = 0;
    /*
     * func:
     * 按照用户查询装备资产，管理员可以查询任意用户，其他用户只能查询自己的资产
     * @param: userid: 管理员可以指定user进行查询
     */
    virtual void getAssetsByUser() = 0;
    /*
     * func: 系统新生成的新装备，发放给特定用户，只能由管理员调用
     * @param: type_id: 游戏装备类型id
     * @param: assetid:
     * 游戏装备唯一id(先从外部获取装备id,也可以实现成一个自增计数器)
     * @param: userid: 获得游戏装备的用户
     */
    virtual void newAssetToUser() = 0;
    /*
     * func: 玩家将自己的装备交易给其他用户, 默认装备持有者是交易发起者
     * @param: to: 装备接收者
     * @param: assetid: 装备id
     */
    virtual void tradeAsset() = 0;
};

struct GameDemo : public GameAssets, public xchain::Contract {
public:
    const std::string ASSETTYPE = "AssetType_";
    const std::string USERASSET = "UserAsset_";
    const std::string ASSET2USER = "Asset2User_";

    void initialize() {
        xchain::Context* ctx = this->context();
        const std::string& admin = ctx->arg("admin");
        if (admin.empty()) {
            ctx->error("missing admin address");
            return;
        }

        ctx->put_object("admin", admin);
        ctx->ok("initialize success");
    }

    bool isAdmin(xchain::Context* ctx, const std::string& caller) {
        std::string admin;
        if (!ctx->get_object("admin", &admin)) {
            return false;
        }
        return (admin == caller);
    }

    void addAssetType() {
        xchain::Context* ctx = this->context();
        const std::string& caller = ctx->initiator();
        if (caller.empty()) {
            ctx->error("missing initiator");
            return;
        }

        if (!isAdmin(ctx, caller)) {
            ctx->error("you do not have permission to call this method");
            return;
        }

        const std::string& typeId = ctx->arg("type_id");
        if (typeId.empty()) {
            ctx->error("missing 'type_id' as asset type identity");
            return;
        }

        const std::string& typeDesc = ctx->arg("type_desc");
        if (typeDesc.empty()) {
            ctx->error("missing type_desc");
            return;
        }

        std::string assetTypeKey = ASSETTYPE + typeId;
        std::string value;
        if (ctx->get_object(assetTypeKey, &value)) {
            ctx->error("the type_id is already exist, please check again");
            return;
        }
        ctx->put_object(assetTypeKey, typeDesc);
        ctx->ok(typeId);
    }

    void listAssetType() {
        xchain::Context* ctx = this->context();
        std::unique_ptr<xchain::Iterator> iter =
            ctx->new_iterator(ASSETTYPE, ASSETTYPE + "~");
        std::string result;
        while (iter->next()) {
            std::pair<std::string, std::string> res;
            iter->get(&res);
            if (res.first.length() > ASSETTYPE.length()) {
                result += res.first.substr(ASSETTYPE.length()) + ":" +
                          res.second + '\n';
            }
        }
        ctx->ok(result);
    }

    void getAssetsByUser() {
        xchain::Context* ctx = this->context();
        const std::string& caller = ctx->initiator();
        if (caller.empty()) {
            ctx->error("missing initiator");
            return;
        }
        std::string userId = caller;
        if (isAdmin(ctx, caller)) {
            // admin can get the asset data of other users
            const std::string& userId2 = ctx->arg("user_id");
            if (!userId2.empty()) {
                userId = userId2;
            }
        }

        std::string userAssetKey = USERASSET + userId + "_";
        std::unique_ptr<xchain::Iterator> iter =
            ctx->new_iterator(userAssetKey, userAssetKey + "~");
        std::string result;
        while (iter->next()) {
            std::pair<std::string, std::string> res;
            iter->get(&res);
            if (res.first.length() > userAssetKey.length()) {
                std::string assetId = res.first.substr(userAssetKey.length());
                std::string typeId = res.second;
                std::string assetTypeKey = ASSETTYPE + typeId;
                std::string assetDesc;
                if (!ctx->get_object(assetTypeKey, &assetDesc)) {
                    // asset type id not found ,skip this asset
                    continue;
                }
                result += "assetid=" + assetId + ",type_id=" + typeId +
                          ",assetDesc=" + assetDesc + '\n';
            }
        }
        ctx->ok(result);
    }

    void newAssetToUser() {
        xchain::Context* ctx = this->context();
        const std::string& caller = ctx->initiator();
        if (caller.empty()) {
            ctx->error("missing initiator");
            return;
        }

        if (!isAdmin(ctx, caller)) {
            ctx->error("you do not have permission to call this method");
            return;
        }

        const std::string& userId = ctx->arg("user_id");
        if (userId.empty()) {
            ctx->error("missing userid");
            return;
        }

        const std::string& typeId = ctx->arg("type_id");
        if (typeId.empty()) {
            ctx->error("missing type_id");
            return;
        }
        std::string assetTypeKey = ASSETTYPE + typeId;
        std::string value;

        std::stringstream ss;

        ctx->get_object(assetTypeKey, &value);
        if (value.empty()) {
            ss << "asset type " << typeId << " not found";
            ctx->error(ss.str());
            return;
        }

        const std::string& assetId = ctx->arg("asset_id");
        if (assetId.empty()) {
            ctx->error("missing assetid");
            return;
        }

        std::string assetKey = ASSET2USER + assetId;
        std::string tmp;
        if (ctx->get_object(assetKey, &tmp)) {
            ctx->error("the asset id is already exist, please check again");
            return;
        }

        std::string userAssetKey = USERASSET + userId + "_" + assetId;
        if (!ctx->put_object(userAssetKey, typeId) ||
            !ctx->put_object(assetKey, userId)) {
            ctx->error("failed to generate asset to user");
        }
        ctx->ok(assetId);
    }

    void tradeAsset() {
        xchain::Context* ctx = this->context();
        const std::string& from = ctx->initiator();
        if (from.empty()) {
            ctx->error("missing initiator");
            return;
        }

        const std::string& to = ctx->arg("to");
        if (to.empty()) {
            ctx->error("missing to");
            return;
        }

        const std::string& assetId = ctx->arg("asset_id");
        if (assetId.empty()) {
            ctx->error("missing assetid");
            return;
        }
        std::string userAssetKey = USERASSET + from + "_" + assetId;
        std::string assetType;
        if (!ctx->get_object(userAssetKey, &assetType)) {
            ctx->error("you don't have assetid:" + assetId);
            return;
        }

        if (!ctx->delete_object(userAssetKey)) {
            ctx->error("failed to delete assetid:" + assetId);
            return;
        }

        std::string assetKey = ASSET2USER + assetId;
        std::string newUserAssetKey = USERASSET + to + "_" + assetId;
        if (!ctx->put_object(newUserAssetKey, assetType) ||
            !ctx->put_object(assetKey, to)) {
            ctx->error("failed to save assetid:" + assetId);
            return;
        }
        ctx->ok(assetId);
    };
};

DEFINE_METHOD(GameDemo, initialize) { self.initialize(); }

DEFINE_METHOD(GameDemo, addAssetType) { self.addAssetType(); }

DEFINE_METHOD(GameDemo, listAssetType) { self.listAssetType(); }

DEFINE_METHOD(GameDemo, getAssetsByUser) { self.getAssetsByUser(); }

DEFINE_METHOD(GameDemo, newAssetToUser) { self.newAssetToUser(); }

DEFINE_METHOD(GameDemo, tradeAsset) { self.tradeAsset(); }